File: src/main/java/engine/rendering/instances/renderers/pbr/PBRRenderer.java
package engine.rendering.instances.renderers.pbr;

import engine.architecture.models.Model;
import engine.architecture.scene.SceneContext;
import engine.architecture.scene.entity.Entity;
import engine.architecture.scene.node.Node;
import engine.rendering.abstracted.renderers.Renderer3D;
import engine.utils.libraryBindings.maths.joml.FrustumIntersection;
import engine.utils.libraryBindings.maths.joml.Vector3f;
import engine.utils.libraryBindings.maths.utils.Matrix4;
import engine.utils.libraryBindings.opengl.shaders.*;
import engine.utils.libraryBindings.opengl.textures.ITexture;
import engine.utils.libraryBindings.opengl.utils.GlUtils;
import lombok.Getter;

public class PBRRenderer extends Renderer3D<PBRModel> {

    private final ShadersProgram<PBRModel> shadersProgram;
    private final FrustumIntersection frustumIntersection = new FrustumIntersection();

    private static final String VERT_FILE = "res/shaders/pbr/pbr_vs.glsl";
    private static final String FRAG_FILE = "res/shaders/pbr/pbr_fs.glsl";

    @Getter
    private static final PBRRenderer instance;

    static {
        try {
            instance = new PBRRenderer();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private PBRRenderer() throws Exception {
        this.shadersProgram = ShadersProgram.create(VERT_FILE, FRAG_FILE);

        shadersProgram.addPerInstanceUniform(new UniformBooleanProperty<>("map_albedo") {
            @Override
            public boolean getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().isAlbedoMapped();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformTextureProperty<>("albedoMap", 0) {
            @Override
            public ITexture getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().getAlbedoMap();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformValueProperty<>("albedoConst") {
            @Override
            public Vector3f getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().getAlbedoConst();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformBooleanProperty<>("map_normal") {
            @Override
            public boolean getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().isNormalMapped();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformTextureProperty<>("normalMap", 1) {
            @Override
            public ITexture getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().getNormalMap();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformBooleanProperty<>("map_roughness") {
            @Override
            public boolean getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().isRoughnessMapped();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformTextureProperty<>("roughnessMap", 2) {
            @Override
            public ITexture getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().getRoughnessMap();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformFloatProperty<>("roughnessConst") {
            @Override
            public float getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().getRoughnessConst();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformBooleanProperty<>("map_metal") {
            @Override
            public boolean getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().isMetalMapped();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformTextureProperty<>("metalMap", 3) {
            @Override
            public ITexture getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().getMetalMap();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformFloatProperty<>("metalConst") {
            @Override
            public float getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getMaterial().getMetalConst();
            }
        });

        shadersProgram.addPerRenderUniform(new UniformValueProperty<>("projectionMatrix") {
            @Override
            public UniformValue getUniformValue(RenderState<PBRModel> state) {
                return getContext().getCamera().getProjectionMatrix();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformValueProperty<>("modelMatrix") {
            @Override
            public UniformValue getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getTransform().getTransformationMatrix();
            }
        });

        shadersProgram.addPerRenderUniform(new UniformValueProperty<>("viewMatrix") {
            @Override
            public UniformValue getUniformValue(RenderState<PBRModel> state) {
                return getContext().getCamera().getViewMatrix();
            }
        });

        shadersProgram.addPerInstanceUniform(new UniformFloatProperty<>("UVscale") {
            @Override
            public float getUniformValue(RenderState<PBRModel> state) {
                return state.getInstance().getUVscalar();
            }
        });
    }

    @Override
    public void render(SceneContext context) {
        if (renderList.isEmpty()) return;

        shadersProgram.bind();
        GlUtils.disableBlending();

        RenderState<PBRModel> renderState = new RenderState<>(this, context.getCamera());
        shadersProgram.updatePerRenderUniforms(renderState);

        for (Model model : renderList.keySet()) {
            for (int i = 0; i < model.getMeshes().length; i++) {
                model.bindAndConfigure(i);

                for (PBRModel entity : renderList.get(model)) {
                    frustumIntersection.set(
                            context.getCamera().getProjectionViewMatrix().mul(
                                    entity.getTransform().getTransformationMatrix(), Matrix4.pool.poolAndGive()
                            )
                    );
                    if (!checkRenderPass(entity)) continue;

                    RenderState<PBRModel> instanceState = new RenderState<>(this, entity, context.getCamera(), i);
                    shadersProgram.updatePerInstanceUniforms(instanceState);
                    model.render(instanceState, i);
                }
            }
        }
        shadersProgram.unbind();
    }

    @Override
    public void render(SceneContext context, Node.Condition condition) {
        if (renderList.isEmpty()) return;

        shadersProgram.bind();
        RenderState<PBRModel> renderState = new RenderState<>(this, context.getCamera());
        shadersProgram.updatePerRenderUniforms(renderState);

        for (Model model : renderList.keySet()) {
            for (int i = 0; i < model.getMeshes().length; i++) {
                model.bindAndConfigure(i);

                for (PBRModel entity : renderList.get(model)) {
                    frustumIntersection.set(
                            context.getCamera().getProjectionViewMatrix().mul(
                                    entity.getTransform().getTransformationMatrix(), Matrix4.pool.poolAndGive()
                            )
                    );
                    if (!checkRenderPass(entity)) continue;

                    RenderState<PBRModel> instanceState = new RenderState<>(this, entity, context.getCamera(), i);
                    shadersProgram.updatePerInstanceUniforms(instanceState);
                    model.render(instanceState, i);
                }
            }
        }
        shadersProgram.unbind();
    }

    private boolean checkRenderPass(Entity entity) {
        return !checkClippingCulling(entity.getTransform().getPosition()) &&
                frustumIntersection.testAab(entity.getModel().getBounds().getMin(), entity.getModel().getBounds().getMax());
    }

    @Override
    public void cleanUp() {
        shadersProgram.delete();
    }
}

File: src/main/java/engine/rendering/instances/renderers/pbr/PBRRenderer.kt
package engine.rendering.instances.renderers.pbr

import engine.architecture.scene.SceneContext
import engine.architecture.scene.entity.Entity
import engine.architecture.scene.node.Node
import engine.rendering.abstracted.renderers.Renderer3D
import engine.utils.libraryBindings.maths.joml.FrustumIntersection
import engine.utils.libraryBindings.maths.joml.Vector3f
import engine.utils.libraryBindings.maths.utils.Matrix4
import engine.utils.libraryBindings.opengl.shaders.*
import engine.utils.libraryBindings.opengl.textures.TextureObject
import engine.utils.libraryBindings.opengl.utils.GlUtils

internal class PBRRenderer : Renderer3D<PBRModel>() {

    private val shadersProgram: ShadersProgram<PBRModel>
    private val frustumIntersection: FrustumIntersection = FrustumIntersection()


    init {
        this.shadersProgram = ShadersProgram.create(VERT_FILE, FRAG_FILE)

        shadersProgram.addPerInstanceUniform(object : UniformBooleanProperty<PBRModel>("map_albedo") {
            override fun getUniformValue(state: RenderState<PBRModel>): Boolean {
                return state.instance.material.isAlbedoMapped
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformTextureProperty<PBRModel>("albedoMap", 0) {
            override fun getUniformValue(state: RenderState<PBRModel>): TextureObject {
                return state.instance.material.getAlbedoMap()
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformValueProperty<PBRModel>("albedoConst") {
            override fun getUniformValue(state: RenderState<PBRModel>): Vector3f {
                return state.instance.material.albedoConst
            }
        })


        shadersProgram.addPerInstanceUniform(object : UniformBooleanProperty<PBRModel>("map_normal") {
            override fun getUniformValue(state: RenderState<PBRModel>): Boolean {
                return state.instance.material.isNormalMapped
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformTextureProperty<PBRModel>("normalMap", 1) {
            override fun getUniformValue(state: RenderState<PBRModel>): TextureObject {
                return state.instance.material.getNormalMap()
            }
        })


        shadersProgram.addPerInstanceUniform(object : UniformBooleanProperty<PBRModel>("map_roughness") {
            override fun getUniformValue(state: RenderState<PBRModel>): Boolean {
                return state.instance.material.isRoughnessMapped
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformTextureProperty<PBRModel>("roughnessMap", 2) {
            override fun getUniformValue(state: RenderState<PBRModel>): TextureObject {
                return state.instance.material.getRoughnessMap()
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformFloatProperty<PBRModel>("roughnessConst") {
            override fun getUniformValue(state: RenderState<PBRModel>): Float {
                return state.instance.material.roughnessConst
            }
        })


        shadersProgram.addPerInstanceUniform(object : UniformBooleanProperty<PBRModel>("map_metal") {
            override fun getUniformValue(state: RenderState<PBRModel>): Boolean {
                return state.instance.material.isMetalMapped
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformTextureProperty<PBRModel>("metalMap", 3) {
            override fun getUniformValue(state: RenderState<PBRModel>): TextureObject {
                return state.instance.material.getMetalMap()
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformFloatProperty<PBRModel>("metalConst") {
            override fun getUniformValue(state: RenderState<PBRModel>): Float {
                return state.instance.material.metalConst
            }
        })




        shadersProgram.addPerRenderUniform(object : UniformValueProperty<PBRModel>("projectionMatrix") {
            override fun getUniformValue(state: RenderState<PBRModel>): UniformValue {
                return getContext().camera.projectionMatrix
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformValueProperty<PBRModel>("modelMatrix") {
            override fun getUniformValue(state: RenderState<PBRModel>): UniformValue {
                return state.instance.transform.transformationMatrix
            }
        })
        shadersProgram.addPerRenderUniform(object : UniformValueProperty<PBRModel>("viewMatrix") {
            override fun getUniformValue(state: RenderState<PBRModel>): UniformValue {
                return getContext().camera.viewMatrix
            }
        })
        shadersProgram.addPerInstanceUniform(object : UniformFloatProperty<PBRModel>("UVscale") {
            override fun getUniformValue(state: RenderState<PBRModel>): Float {
                return state.instance.uVscalar
            }
        })
    }

    override fun render(context: SceneContext) {
        if (renderList.size < 1) {
            return
        }
        shadersProgram.bind()

        GlUtils.disableBlending()

        val renderState = RenderState<PBRModel>(this, context.camera)
        shadersProgram.updatePerRenderUniforms(renderState)

        for (model in renderList.keys) {
            for (i in model.meshes.indices) {
                model.bindAndConfigure(i)
                for (entity in renderList[model]!!) {
                    frustumIntersection.set(
                        context.camera.projectionViewMatrix.mul(
                            entity.transform.transformationMatrix, Matrix4.pool.poolAndGive()
                        )
                    )
                    if (!checkRenderPass(entity) && entity.isActivated) {
                        continue
                    }
                    val instanceState = RenderState<PBRModel>(this, entity, context.camera, i)
                    shadersProgram.updatePerInstanceUniforms(instanceState)
                    model.render(instanceState, i)
                }
            }
        }

        shadersProgram.unbind()
    }

    override fun render(context: SceneContext, condition: Node.Condition) {
        if (renderList.size < 1) {
            return
        }
        shadersProgram.bind()

        val renderState = RenderState<PBRModel>(this, context.camera)
        shadersProgram.updatePerRenderUniforms(renderState)

        for (model in renderList.keys) {
            for (i in 0..model.meshes.size) {
                model.bindAndConfigure(i)
                for (entity in renderList[model]!!) {
                    frustumIntersection.set(
                        context.camera.projectionViewMatrix.mul(
                            entity.transform.transformationMatrix, Matrix4.pool.poolAndGive()
                        )
                    )
                    if (!checkRenderPass(entity) && entity.isActivated) {
                        continue
                    }
                    val instanceState = RenderState<PBRModel>(this, entity, context.camera, i)
                    shadersProgram.updatePerInstanceUniforms(instanceState)
                    model.render(instanceState, i)
                }
            }
        }

        shadersProgram.unbind()
    }

    private fun checkRenderPass(entity: Entity): Boolean {
        return !checkClippingCulling(entity.transform.position) &&
                frustumIntersection.testAab(entity.model.bounds.min, entity.model.bounds.max)
    }

    override fun cleanUp() {
        shadersProgram.delete()
    }

    companion object {

        private const val VERT_FILE = "res/shaders/pbr/pbr_vs.glsl"
        private const val FRAG_FILE = "res/shaders/pbr/pbr_fs.glsl"

        @JvmStatic
        val instance: PBRRenderer = PBRRenderer()

    }
}
